### ISSUE TITLE
Improve docs for locale-default calendar (was: Does locale have default calendar?)

### ISSUE THREAD

**Issue by rizzen-yazston** [unicode-org/icu4x#3008]
I am busy implementing the DateTimeFormatter or DateFormatter into my i18n library crate, have it working with the non-experimental options.

From the examples of the docs for both DateTimeFormatter and DateFormatter, I have noted the inclusion of the `-u-ca-?` tag in creating the Locale object. Where ? represents any of the 8 calendars.

I am wondering if locales have a default calendar, like `en` uses `gregorian`, `jp` uses `japanese`, or must the calendar be explicit set in the locale using the `-u-ca-?` tag, whenever one wants to format a date using a DateTime or Date object?

Documentation is vague regarding default calendar.

For time and date the documentation mentions that the lengths to be set to `medium` when display options are not present.

**Comment by rizzen-yazston** [comment]
Would it be reasonable to make Iso calendar the default for any locale?

**Comment by sffc** [comment]
If you use DateFormatter or DateTimeFormatter, the calendar is automatically chosen for the locale. According to the current CLDR release most locales default to `gregory` except a few that default to `buddhist`.

Seems like an area where docs could be improved.

**Comment by rizzen-yazston** [comment]
Thanks for the prompt reply.

Correct to say that one only needs to add the Unicode extension calendar tag `-u-ca-?` to change from the locale's calendar default?

Current examples to me seem to indicate that there is no default calendar for locales.

**Comment by sffc** [comment]
Right, `-u-ca` allows the user to override their locale's default calendar.

**Comment by sffc** [comment]
Hi @rizzen-yazston, could you peruse the latest docs, which were rewritten in the 2.0 release, and give feedback on whether they are more clear?

https://unicode-org.github.io/icu4x/rustdoc/icu/datetime/index.html

Also note the new DateTimeFormatterPreferences struct that allows setting the calendar via an enum

https://unicode-org.github.io/icu4x/rustdoc/icu/datetime/struct.DateTimeFormatterPreferences.html

**Comment by sffc** [comment]
Hi @rizzen-yazston, still would be interested in your feedback!

**Comment by rizzen-yazston** [comment]
Hi @sffc, I apologise for the late response. I have been busy trying to finalise another project for a GUI library, so I have no time to look at the latest ICU4X developments. I see that some additional features have been added since I created my I18N library using ICU4X, so I will work on the library to incorporate the improvements of ICU4X.

**Comment by rizzen-yazston** [comment]
After a quick scan of the documentation, while also reviewing my library's source (using ICU4X 1.5.0), I am identifying where the changes are being made, and comparing the `DateTimeFormatter` implementations of version 2.2.1 and 1.5.0. The examples provided for existing and new functions are clearer and now enable easy retrieval of the default calendar for a locale (or `LanguageTag`), as well as easier modification of the locale's calendar using the new `DateTimeFormatterPreferences` struct and the provided functions.

From the documentation alone, there is a noticeable improvement in the documentation and the various date-time structs.

I'll let you know if I come across any issues while updating my library to use ICU4X 2.2.1.

**Comment by rizzen-yazston** [comment]
Looking at `DateTimeFormatter::try_new()`, I have noticed that the field set specifies "styling" (long, medium, short) for dates and times. The locale is now embedded within `LocalePreferences` of the `DateTimeFormatterPreferences` struct.

Does each locale have a default/preferred field set? If so, how does one retrieve this field set?

Am I correct in understanding that if one wants to support any "styling" of dates and times, one needs to use the all-inclusive enum?

**Comment by sffc** [comment]
Hi @rizzen-yazston!

I think we've tried to avoid using the word "style", since that term is overloaded with meaning from older versions of ICU. We use the term "length" instead.

The field set is not a locale preference. However, how the combination of field set and length map to specific field lengths is locale-specific. For example, a field set YMD with length Short might map to the number 1 in English but a string in another language that doesn't use month numbers.

> Am I correct in understanding that if one wants to support any "styling" of dates and times, one needs to use the all-inclusive enum?

ICU4X 2.0 introduced field sets with options that together control the styling of dates and times.

If there's room to improve the docs, I'd love to hear it.

**Comment by rizzen-yazston** [comment]
Busy implementing the new structs, and it is definitely a paradigm shift in how dates and times are formatted. I discovered that I need to add an optional date-time fieldset to the pattern specifications. I do see the benefit of having a fieldset within the pattern.

However, if not specified, I would like you to use the default fieldset for the locale if no fieldset is specified, thus how does one obtain the default fieldset from the locale? Else must the default be an application-specified default?

**Comment by sffc** [comment]
I see. So, there's never actually been CLDR data that makes the set of fields change between locales. The "style" formats we used in ICU4X 1.5 had been previously implemented by ICU4C for 20+ years, and through all that time, there was never a locale that use a different field set, only different field lengths. So the new design adopted by CLDR and implemented by ICU4X is to pull out the field set from the locale, and let the client specify it, since it most often depends on application context.

For example, to determine whether to say "January 12" instead of "January 12, 2026", you need to know whether the application makes it clear from context that the date is in the same year. That isn't something that the locale can determine.

**Comment by rizzen-yazston** [comment]
Interesting.

Perhaps it would be a good idea, on the `fieldsets` module documentation, to have a notice statement that there are no preferred (default) fieldsets per locale (language and region) in the CLDR. If preferred fieldsets are required, the application must provide the preferred (default) fieldset per context.

**Comment by rizzen-yazston** [comment]
I have noticed in the `fieldsets` modules that each date-time component has a unique letter assigned. What is the reason for having many letter combination structs, instead of a single struct that takes a `&str` parameter containing the string of selected characters (duplicates and unknown letters are simply ignored)?

**Comment by rizzen-yazston** [comment]
I have noticed that the documentation at `https://unicode.org/reports/tr35/#UnicodeCalendarIdentifier` lists calendar names as `islamic`, while `icu_calendar::preferences::CalendarAlgorithm` has variant `Hijri` and `icu_calendar::preferences::HijriCalendarAlgorithm` includes variants of those listed in the report TR35.

**Comment by rizzen-yazston** [comment]
Function `try_new()` on `DateTimeFormatter` is missing the requirement of the crate feature `compiled_data` in the function's description.

**Comment by rizzen-yazston** [comment]
Working with the new fieldsets is not so trivial, but I managed to get a working solution; however, it became rather bloated because I could not create a function that would use the correct "try_new" function and could accept any of the fieldset structs.

I ended up with this:

```
                    let date_string = match localiser.icu_data_provider().as_ref() {
                        #[cfg(feature = "icu_compiled_data")]
                        IcuDataProvider::Internal => match field_set_string.as_str() {
                            "D" => {
                                DateTimeFormatter::try_new(prefs, fieldsets::D::for_length(*date))?
                                    .format(&iso_date_time)
                                    .to_string()
                            }
                            "DE" => {
                                DateTimeFormatter::try_new(prefs, fieldsets::DE::for_length(*date))?
                                    .format(&iso_date_time)
                                    .to_string()
                            }


   other fieldsets


                            _ => {
                                // Should never reach.
                                return Err(FormatterError::NeverReached);
                            }
                        },

                        #[cfg(feature = "blob")]
                        IcuDataProvider::Blob(provider, _) => match field_set_string.as_str() {
                            "D" => DateTimeFormatter::try_new_with_length_with_buffer_provider(
                                provider,
                                prefs,
                                fieldsets::D::for_length(*date),
                            )?
                            .format(&iso_date_time)
                            .to_string(),
                            "DE" => DateTimeFormatter::try_new_with_length_with_buffer_provider(
                                provider,
                                prefs,
                                fieldsets::DE::for_length(*date),
                            )?
                            .format(&iso_date_time)
                            .to_string(),


   other fieldsets


                            _ => {
                                // Should never reach.
                                return Err(FormatterError::NeverReached);
                            }
                        },


other provider branches


                        #[allow(unreachable_patterns)]
                        _ => return Err(FormatterError::NoIcuProvider),
                    };

```

Is there a way to simplify the process of removing duplicate blocks to obtain the formatting? Trying to sort out the `FSet` requirements is driving me nuts.

**Comment by sffc** [comment]
I really appreciate your time working with this code.

It looks to me like you should consider using [fieldsets::enums](https://unicode-org.github.io/icu4x/rustdoc/icu_datetime/fieldsets/enums/index.html) which allows you to write an enum-based dispatch, or perhaps [fieldsets::builder](https://unicode-org.github.io/icu4x/rustdoc/icu_datetime/fieldsets/builder/index.html) which is a fully dynamic approach, more like 1.5.

We made the new trait-based fieldset approach the default in 2.0 because it delivers the best binary size and performance, but it's still very important that we do a good job documenting the other two approaches (enums and builder) and helping clients understand which solution is the best for their use case. There's a lot of good info here to help improve the docs.

**Comment by rizzen-yazston** [comment]
It is a pleasure. Though my use case tends to be on the more unusual side. 😎 

Thanks for the information. Will look at the two options, though `fieldsets::builder` module might be what I need for my library. Just need to add the time zone support to my patterns.

**Comment by rizzen-yazston** [comment]
I see no time examples for `Length`, only for date examples. I am guessing that `Length` has nothing to do with which time fields are to be included (hour, minute, second, and microseconds). For time fields, one needs to use these functions (`hm()`, `hms()`, and `hmss(...)`). Why is there no `h()` for just the hour?

If one wants to specify both time length and time precision, one needs to construct the struct itself?

**Comment by sffc** [comment]
> I see no time examples for `Length`, only for date examples. I am guessing that `Length` has nothing to do with which time fields are to be included (hour, minute, second, and microseconds).

That's correct. Time length impacts the length of the AM/PM in certain locales. The fields being displayed are not impacted.

> Why is there no `h()` for just the hour?

Not a good reason, just that it seemed less "readable" than hm and hms. You can get it from TimePrecision instead.

> If one wants to specify both time length and time precision, one needs to construct the struct itself?

Yeah, it seemed like customizing the time length was not common, so we didn't optimize for setting the length like we did for the date fields.

**Comment by rizzen-yazston** [comment]
Just letting you know I am still converting my i18n library to use the new date-time structs and enums. I was without my laptop for 2 weeks while it was being repaired.

Made changes to my date-time pattern and updated the syntax grammar. Now busy translating the date time pattern options to ICU4X structs and enums. Adding date-time defaults to the library (Gregorian medium-length month-day and hour-minute time); however, the application will be able to change the defaults.

**Comment by rizzen-yazston** [comment]
I see the `FieldSetBuilder` struct has only one `length` field, which I assume is used for both the date and time components. Is it not possible to have different lengths for the date and time components? If so, the struct would need two length fields. Or do you (or the WG) believe users would never have a need to specify different lengths for date and time components?

**Comment by sffc** [comment]
There are only 4 fields that are influenced by the length: Month, Weekday, Era, and Day Period. (Note: time zone length is controlled by its own option.)

- Era widths other than Short are rarely used; most locales only have data for Short. I have never seen code that tries to use a Long era width while other fields are not also Long. Happy to be shown evidence to the contrary.
- Day Period widths other than Short are also rarely used, except in flexible day periods, which we don't currently support (#487).

So the only real need to have different widths is between Month and Weekday. For example, we don't currently support: "Wednesday, Feb 18" (long weekday, short month) or "Wed, February 18" (short weekday, long month). This is a legitimate use case and is tracked in https://unicode-org.atlassian.net/browse/CLDR-18003

**Comment by rizzen-yazston** [comment]
Looking at the source for `FieldSetBuilder` and `FieldSet`, the field `length` of `FieldSetBuilder` via the `take_from_builder()` is used for both date and time, and `Length::Medium` is used when the `length` is `None`. 

If a length is unsupported for a particular date-time field, is it silently converted to a supported length? This is the case with the era field, which has only `Short` for most locales, with the default length being `Medium`.

Is the field `length` of the time struct `T` just ignored?

To support different lengths for the month and the day of the week, `FieldSetBuilder` would need multiple length fields, and `take_from_builder()` would select the appropriate length(s).

Just wondering, would it not be better to separate the field `Length` from the fieldsets (T, YMD, MDE, etc) and place supported lengths in the `DateTimeFormatterPreferences` to have multiple lengths in the `FieldSetBuilder` struct, and `take_from_builder()` having to select the correct lengths?

**Comment by rizzen-yazston** [comment]
> There are only 4 fields that are influenced by the length: Month, Weekday, Era, and Day Period. (Note: time zone length is controlled by its own option.)
> 
>     * Era widths other than Short are rarely used; most locales only have data for Short. I have never seen code that tries to use a Long era width while other fields are not also Long. Happy to be shown evidence to the contrary.
> 
>     * Day Period widths other than Short are also rarely used, except in flexible day periods, which we don't currently support ([Implement flexible day periods #487](https://github.com/unicode-org/icu4x/issues/487)).
> 
> 
> So the only real need to have different widths is between Month and Weekday. For example, we don't currently support: "Wednesday, Feb 18" (long weekday, short month) or "Wed, February 18" (short weekday, long month). This is a legitimate use case and is tracked in https://unicode-org.atlassian.net/browse/CLDR-18003

To support different lengths for the month and the day of the week, the current fieldsets `ME`, `MDE`, `YMDE`, and `YMDET` would need to be altered to include 2 length fields.

Even though this is not supported in the current ICU4X version, I would like to include all required lengths in my library's pattern specification. Thinking of having a keyword `length`, that translates to the functions `long()`, `medium()`, or `short()` of the fieldsets `MD`, `YMD`, etc. To handle the day-of-week length (future, when the ICU4X fieldsets support both lengths), add another keyword `length_week`. Finally, have a separate keyword `length_zone` to represent the time zone.

Think there is a typo error with the example on the web page [Date/Time Patterns](https://cldr.unicode.org/translation/date-time/date-time-patterns), section "Synchronizing Date/Time Names and Patterns", subsection "When to use Standalone vs. Formatting", of list item "stand-alone style: Use the stand-alone month names represented with “L” when the month name is without a day-of-the month. (e.g. LLLL y: Monday 1999)". Should this not be "LLLL y: January 1999"?

**Comment by rizzen-yazston** [comment]
For fieldset `T`, aren't these `ms`, `mss`, and `ss` also valid time formats, with the hour and/or minute not displayed? Especially for events that never extend into the hour or minute (really short events).

**Comment by sffc** [comment]
`ms`, `mss`, and `ss` are not valid time formats. They are valid durations, but not times.

The way I envision field-specific lengths working is by adding another field to the fieldset options, like `length_month`, as you suggested. It would default to `length` but allows for an override. It just hasn't been implemented yet.

**Comment by rizzen-yazston** [comment]
It means I need a new, separate pattern in my library to handle durations and use the `Duration` struct (currently experimental). Thank you for your explanations and guidance.

**Comment by rizzen-yazston** [comment]
How does one specify the significant fields for the `DurationFormatter`? Like the start and end fields.

For example, in a 100 m sprint, only seconds and milliseconds matter. Year, month, day, hour, minute, microseconds and nanoseconds are not displayed. Only 2 of the 9 fields, excluding the sign field, which, for this example, will always be positive.

For an ultra-long marathon, the significant fields would be hours, minutes and seconds, which are 3 of the 9 fields.

**Comment by rizzen-yazston** [comment]
What are the valid field combinations that include the week field in the `Duration` struct?

What happens when the `week_visibility`, `month_visibility` and `year_visibility` are set to `FieldDisplay::Always` in the `ValidatedDurationFormatterOptions`?

`DurationFormatterPreferences` does not have the field `pub calendar_algorithm: Option<CalendarAlgorithm>,` that `DateTimeFormatterPreferences` does have. Does this mean duration is consistently formatted according to the locale's default calendar?

**Comment by rizzen-yazston** [comment]
Finally managed to sort out creating the `FSet` using the builder method (using `build_composite()` after setting all the required fields), and creating the preference instance needed for `DateTimeFormatter::try_new()` to have a `DateTimeFormatter` instance.

I am struggling to convert a `DateTime<Iso>` value to a type that `.format()` will accept. I look at the various examples, yet I am failing.

**Comment by rizzen-yazston** [comment]
I am definitely missing something important.

I managed to create the date-time formater, but when I pass `date_time_iso` (`&DateTime<Iso>`) to `format()`, I get this error, and I'm struggling to figure out what is missing.

```
let formatted = DateTimeFormatter::try_new(*preference, *composite)?.format(date_time_iso);
```

```
the trait bound `icu_time::DateTime<icu_calendar::Ref<'_, AnyCalendar>>: AllInputMarkers<CompositeFieldSet>` is not satisfied
the following other types implement trait `GetField<T>`:
  `icu_time::DateTime<A>` implements `GetField<()>`
  `icu_time::DateTime<A>` implements `GetField<MonthInfo>`
  `icu_time::DateTime<A>` implements `GetField<Nanosecond>`
  `icu_time::DateTime<A>` implements `GetField<YearInfo>`
  `icu_time::DateTime<A>` implements `GetField<icu_calendar::types::DayOfMonth>`
  `icu_time::DateTime<A>` implements `GetField<icu_calendar::types::DayOfYear>`
  `icu_time::DateTime<A>` implements `GetField<icu_calendar::types::RataDie>`
  `icu_time::DateTime<A>` implements `GetField<icu_calendar::types::Weekday>`
and 3 others
required for `icu_time::DateTime<icu_calendar::Ref<'_, AnyCalendar>>` to implement `AllInputMarkers<CompositeFieldSet>`
```

**Comment by Manishearth** [comment]
I think as far as this issue title is concerned, this issue is fixed for 2.2, based on https://github.com/unicode-org/icu4x/issues/3008#issuecomment-3707243876.

There do seem to be related things that could be improved, and that discussion can continue until we have something actionable.

**Comment by sffc** [comment]
> What are the valid field combinations that include the week field in the `Duration` struct?

Duration formatting doesn't use field sets.

> What happens when the `week_visibility`, `month_visibility` and `year_visibility` are set to `FieldDisplay::Always` in the `ValidatedDurationFormatterOptions`?

For DurationFormatter, we are following the ECMA-402 spec very closely.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DurationFormat

> `DurationFormatterPreferences` does not have the field `pub calendar_algorithm: Option<CalendarAlgorithm>,` that `DateTimeFormatterPreferences` does have. Does this mean duration is consistently formatted according to the locale's default calendar?

Durations are calendar-agnostic.
